### PR TITLE
Add mz_cluster_replica_metrics table

### PR DESCRIPTION
**Pull request by umanwizard**
For each replica, spin up a loop that queries Kubernetes for its metrics (CPU and Memory) every 60 seconds, and dumps them into a system table.

There was some back-and-forth on whether we need to encode the timestamp as part of the data, but it seems unnecessary for now.

### Motivation

  * This PR adds a known-desirable feature.

    Touches https://github.com/MaterializeInc/materialize/issues/15808 . We will need a follow-up PR to do this for storage hosts.
    

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Add `mz_internal.mz_cluster_replica_metrics` system table

### CODE REVIEW

**Review by teskje — APPROVED** [review]
LGTM

It occurred to me that monitoring pod metrics is a similar operation to watching for pod status updates (what `watch_services` is doing). Both are tasks running in the background that periodically return some information about our replica processes from the orchestrator. But we handle them quite differently in this PR. Did you consider treating the metric fetching the same way as `watch_services`, or even making `watch_services` also return metrics updates?

**Review by vmarcos — COMMENTED** [review]
LGTM overall; however, it would be good to address the following points before merging:

- We need user-facing documentation for the new `mz_internal.mz_cluster_replica_metrics` relation being added. I believe that the relevant page to be updated should be https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/sql/system-catalog/mz_internal.md ?
- Jan's point regarding dead code for `MZ_STORAGE_HOST_METRICS`.

I also have a general question: Is there some way you could add a test (e.g., using `mzcompose` or platform checks) where you actually check for obtaining some data out of `mz_internal.mz_cluster_replica_metrics`? Right now, we are only testing that the new relation is in the catalog, so even a test that checks for basic data publication (some well-formed measurements came out, not a retraction error or equivalent) would give us some additional reassurance.

**Comment by umanwizard** [comment]
Answering the raised points:

1. Yes, the storage stuff is dead code now. I will create the mz_storage_host_metrics table in a follow-up PR, unless Gus prefers to do it.
2. No problem, I can add docs,
3. Unfortunately this is not straightforward to test as the metrics server only runs in real clusters. I can check whether it's not too difficult to add it to cloudtest, but if that's hard to do then IMO, testing by hand is an acceptable stopgap.

**Comment by umanwizard** [comment]
4. I didn't know about this k8s `watcher` concept but it's a good idea; I'll try it out

File: src/adapter/src/catalog/builtin.rs [hunk 1]
```diff
@@ -1532,6 +1532,29 @@ pub static MZ_EGRESS_IPS: Lazy<BuiltinTable> = Lazy::new(|| BuiltinTable {
     desc: RelationDesc::empty().with_column("egress_ip", ScalarType::String.nullable(false)),
 });
 
+pub static MZ_CLUSTER_REPLICA_METRICS: Lazy<BuiltinTable> = Lazy::new(|| BuiltinTable {
+    name: "mz_cluster_replica_metrics",
+    // TODO[btv] - make this public once we work out whether and how to fuse it with
+    // the corresponding Storage tables.
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/adapter/src/catalog/builtin.rs:1538]: That is up to the storage team.

`mz_storage_host_metrics` is the only one I know for sure we'll want in the short term.

In the long term, the concept of "storage hosts" and "clusters" is probably being combined into one thing, so we will only need one table.

**Review by benesch — COMMENTED** [review]
Really excited to see this come together. A few thoughts, some substantive and some procedural:

  * Could you loop in the surfaces team (@chaas/@ggnall) when adding new relations to the system catalog, so they can check them for consistency with the other system tables?
  * `nano_cpus` and `bytes_memory` are inconsistent with the rest of the system catalog, which use units as suffixes, rather than prefixes.
  * I think there's a design discussion to be had about whether to use CPU % of limit and memory % of limit rather than absolute CPU/memory. If we stick with absolute CPU/memory than we'll need to surface the limits per cluster size in some other way.

**Comment by umanwizard** [comment]
Thanks for taking a look @benesch . This is already merged but I'll fix the issues you mentioned in a new PR shortly.

**Pull request merged**